### ISSUE TITLE
Proposal: implement typeclass search cahce

### ISSUE THREAD

**Issue by vzaliva** [coq/coq#6213]
Following discussion at the coq-dev mailing list, I am opening this issue to discuss implementing a mechanism to cache typeclass search results to speed up repeated searches.

**Comment by vzaliva** [comment]
from @mattam82  "AFAIR Jason Gross had a proposal in this direction some time ago. In general it is hard because proof search is context-dependent, a query in one context might have a different answer in another. But for an already run (typeclass) resolution in the same context and with the same hints yes, a cache could be kept. I think it would not be too hard to implement on top of the current implementation."

from @JasonGross  "Here's another proposal: in my experience, most of the time spent in typeclass resolution is spent on subtrees that eventually fail.  The cache can simply record which typeclasses goals eventually fail for successful typeclass resolutions, without recording their descendents.  Typeclass resolution using the cache should trust the cache about failures, but if the entire resolution fails, it should try again without the cache.  Since typeclass resolution is fully backtracking, this cannot change the success/failure behavior of type class resolution, though of course the cacheless search can get much slower under your nose without you noticing, so I'd want a setting that disables using the cache."

**Comment by vzaliva** [comment]
I agree with @JasonGross that most of the time following dead branches takes most of the time. But this is just a subjective observation. We can try to analyze some logs to quantify it and see what would be the most optimal way to purge the search tree: positive (guiding towards some branches) or negative (rejecting some branches early).

The positive strategy has an advantage that it could be expressed in terms of current hints mechanism. The cache will become just additional hint database activated during particular search. Not sure if it is easy to reject branches via Hints...

**Comment by vzaliva** [comment]
I plan to try to work on this feature in my spare time. I am new to Coq internals so, I will appreciate some guidance. Few initial questions I have:

1. How the mechanism will be exposed to the user? Should it be a global setting which user just activates (e.g. `Set Typeclass Caching`), or perhaps it should be a command which wraps particular tactics application? (e.g. `cached setoid_rewrite`)
1. How the implementation should be organized (code-wise)? Should just extend existing typeclass search mechanism? One idea is to have an internal API for plugging external search/cache implementations and this to be the first one to use this. There might be possible to create others.
1. How the cache is be stored? We will have the cache for multiple queries. Perhaps a single cache file per .v?
1. Should we try to use some existing mechanism to implement cache application (e.g. temporary transient hints databases which are activated if cache has chosen to be valid)?
1. What is cache invalidation criteria? What should be included in the check upon which cache should be invalidated? Current context? State of hint databases? Anything else? How must these conditions be checked? Perhaps they are invariant of some things like order or names of parameters?
1. Related to the previous questions: What would be the cache file format?
1. If cache validity conditions are satisfied, should we rely on the cache, or there are the situations when we try redo the searches if cached response fails?
1. Finally, what do we cache? Successful searches or rejected braches ( @JasonGross  suggestion)?

Of course some of these questions we will have to figure out as we proceed with the implementation, but it would be great to hear some early thoughts on these.

**Comment by JasonGross** [comment]
1. It should be a setting.  It is a performance optimization, and may need to be disabled for debugging.  If I write a tactic script that invokes setoid_rewrite, it has no way of knowing at Ltac definition time whether to cache or not.
2. I'm not the person to ask, this is a placeholder so GitHub doesn't renumber my list
3. I'd say follow the format of the .aux file, which is already a sort of cache.  Perhaps even store the cache in the .aux file (I guess @gares is the person to ask about this?)
4. *shrugs*  I can't think of a good one
5. At the very least, the type of the goal.  Maybe also the context.  Beware, if you store the entire hint-db, your cache may get very large.  Perhaps it's best only to store a hash of the type of the goal, the context, and the list of types and priorities of hints in the typeclass_intances database (and list unsoundness in the presence of `intuition` as a tc hint as a known bug).
6. I think one important thing here is file size, and one way to reduce it is to replace objects with hashes when you can
7. This is tricky.  Do you want the cache validity condition to be absolutely sound?  Then you'll need to track things like universe constraints and every single hint db if you use `intuition` or `auto with *` in an extern hint, and your cache may be almost always invalid.  I think the better choice may be to document locations of unsoundness, have the cache off by default, and document loudly that this will help speed-wise at the cost of missing corner cases.  I feel a bit iffy about this.
8. I think failures are more robust, even if successes lead to slightly smaller caches.  I may be wrong, and, in any case, I expect this will separate cleanly from the rest of the implementation and it will be easy to test both.

**Comment by vzaliva** [comment]
@JasonGross thanks for your comments. I will be looking at current implementation and may have additional remarks. Another quick question for devs -- what branch/version should I use as starting point to develop this? Is it OK to fork 8.7 or should I work from dev. branch?  

@mattam82 any hints or guidance on implementation are appreciated. E.g. where source code wise you may think this should fit....

**Comment by herbelin** [comment]
@vzaliva: Reacting only to the dev vs 8.7 question: The dev branch is changing so rapidly these days that I would try starting from master to minimize API changes, or even deprecation warnings.

**Comment by mattam82** [comment]
1. Global option I think too
2. It should fit in tactics/class_tactics.ml in the "new resolution code" in Module Search around line 1000. That's where the typeclasses eauto proof search tactic is defined. I'm not sure what kind of API you're thinking about. But somehow, if you could make caching a kind of action of a (state or approaching) monad then we could functorize indeed by this monad implementation to get caching and non-caching versions.
3 and 5. That I don't know what's best. At first you can try a simple minded reference (see Summary.ref for persistent references across Undo)  and a cache invalidation check like "is it the same evar_map (set of goals) modulo renaming of evars" and the same hint database (see hint_db in proofs/hints.mli, we might need to add a structural equality check on those) ? You'll certainly have to study the engine/evd.mli interface a bit to understand the proof state, and proofview.mli as well.
4. It sounds too complicated at first
6. Serializing it does not sound nice (for the positive success at least). For failed branches it is much easier as there is already a type of "search paths" corresponding to it, that's just a list of names of hints.
7. Hmm, tricky one. Only if cache validation is perfect we shall not need to retry the proof, to remain observationally equivalent, no?
8. Both are interesting indeed... I don't know yet

**Comment by vzaliva** [comment]
I did some experiments and have prototyped something illustrating a possible approach to this problem. Before I proceed to implement cache persistence and other stuff it would be great if somebody could take a quick look at let me know that I am on the right path here. The code is in https://github.com/vzaliva/coq/blob/v8.7-typeclass-cache/tactics/class_tactics.ml Look in the module `Search` for `tc_cache_entry` and `tc_cache_entry_cmp` and how they are used in `hints_tac_gl`. I tested it on some simple cases and it seems to be working.

**Comment by ejgallego** [comment]
I think you want to rebase to `master` ASAP; the delta between 8.7 and 8.9 will not be insignificant.

**Comment by JasonGross** [comment]
> Additionally it is not possible to define a hash function to use a hash set.

Why not?

**Comment by vzaliva** [comment]
The main problem with defining hash function is how consts are compared. My implementation is `Universes.eq_constr_univs_evars_infer_with` which is attempts to "unify" two goal's evars based on their types, so goals, say, `T ?a` and `T ?b` will match as long as `?a` and `?b` have the same type.

I was considering re-naming all evars according to same scheme which will give a "canonic" representation of the goal for which then the hash function could be calculated. But such renaming will require making a copy of the goal and will be quite heavy computationally, so I am not sure if it would help.

**Comment by gares** [comment]
I arrive late here... In logic programming "caching" is called tabling, and iirc B.Pientka studied tabling in the higher order case. I m talking about run time cache, not the persistence part

**Comment by CohenCyril** [comment]
Oh, I arrive late here too. I am also interested by this feature!

> Global option I think too

Please don't... No more global settings for typeclass behaviour please (except as a global killswitch). It should be by pattern, so that one could put a tag (e.g. a boolean) on his class: e.g. `Typeclasses Cache (my_class true _ _ _)` to be selective and `Typeclasses Cache (my_class _ _ _ _)` to extend caching to the whole class. There should be attribute to abbreviate the latter.

**Comment by CohenCyril** [comment]
@vzaliva such mechanisms are under the hood unless they do not scale... This mechanisms is a trade-off: you trade space for time. There is a variety of logic programs you could write using typeclasses, independent of each other, in different libraries, and combine them in the same proof, or even in the same proof search. Some may be very redundant (e.g. math-comp), and some may not (an arbitrary proof search), if you do not let the developer of each library decide which search uses caching and which should absolutely not, the burden falls onto the user to set and unset the flag accordingly, or suffer from the incompatibility between two libraries. So, both for the sake of separating library development from library usage and of independence of programs in the library, this must not happen.

**Comment by ejgallego** [comment]
What do other systems do? Does Haskell implement tabling for TC search?

While indeed tabling is usually specified in the way Cyril points out, I am not sure TC is the right place to expose it. If we have a general-purpose logic programming engine in Coq let's expose it properly.

**Comment by CohenCyril** [comment]
> While indeed tabling is usually specified in the way Cyril points out, I am not sure TC is the right place to expose it. If we have a general-purpose logic programming engine in Coq let's expose it properly.

I did not understand these two sentences, can you expand? What, but TC, as of today, would be the general-purpose logic programming framework in Coq? What do you mean by TC is not the right place to expose tabling properly?

**Comment by Zimmi48** [comment]
>What do other systems do? Does Haskell implement tabling for TC search?

I heard that Haskell TC search is much less powerful than Coq's. I can't expand on this because I'm not even sure that it's true. @mattam82 can probably comment.

>While indeed tabling is usually specified in the way Cyril points out, I am not sure TC is the right place to expose it. If we have a general-purpose logic programming engine in Coq let's expose it properly.

>I did not understand these two sentences, can you expand? What, but TC, as of today, would be the general-purpose logic programming framework in Coq? What do you mean by TC is not the right place to expose tabling properly?

I don't know if this is in line with what @ejgallego was trying to convey, but in "Coq's Prolog and application to defining semi-automatic tactics" (https://hal.archives-ouvertes.fr/hal-01671994) I argued that the type class mechanism is being abused while the real logic programming engine is the infrastructure that is behind it and that is accessible by other means. As part of https://github.com/coq/coq/projects/11 / #721, I have been planning to expose this engine more conveniently to the user. It was also my intention to reorganize the doc and the name of the options to highlight this engine independently of the type class mechanism that is built on top. Due to other pressing matters, this will unfortunately never be finished in time for 8.9.

**Comment by vzaliva** [comment]
One very valid point @CohenCyril makes is that unbounded cache growth can cause a memory exhaustion. This certainly should be avoided under any circumstances and I will add a mechanism bounding the cache size.

I also see why people may want finer granular cache control. I will propose some additional options here before implementing. 

But before spending more time on adding these options I want to make sure the caching implementation I propose is something feasible and could be potentially become part of Coq. To do so I suggest first to run standard Coq benchmark to confirm the performance gains and to make sure it does not break anything. I do not have access to Coq's continuous integration suite, but if somebody could work with me to help to run my caching implementation through all the libraries we have in it, it would be great.

I want to remind everybody that my unscientific benchmark show **20% improvement** in compilation speed. I think it is very significant and if confirmed could be a great benefit for Coq users. I would **love to see it in 8.9**.

Also at this point, I suggest to limit the implementation to in-memory caching (without persistence). If it could give significant performance improvements I would rather have it integrated first and work on persistence as a separate improvement, for the next iteration.

**Comment by CohenCyril** [comment]
> I do not have access to Coq's continuous integration suite, but if somebody could work with me to help to run my caching implementation through all the libraries we have in it, it would be great.

If you open a PR from your branch to coq/master, your code will automatically run through Coq CI.

**Comment by ejgallego** [comment]
Indeed, thanks @Zimmi48 , as I understand the situation there are other users of the "LP" system such as waiting, but I am not an expert on this area.

**Comment by Zimmi48** [comment]
>I would love to see it in 8.9.

Then you should urgently open a PR (as was also recommended by @CohenCyril). I can't promise anything to you however as 8.9 branching is quite soon and this patch will require careful review (and most developers are on holidays during part of August).

**Comment by vzaliva** [comment]
Here is a comment by @ppedrot from pull-request:

> Unluckily, I think that this cache mechanism is completely unsound. First, there is no guarantee that unification is stable by the provided equivalence of states. Due to bugs, in general it is not, e.g. when the result depends on some fixed name. But even without taking these bugs into consideration, I don't know what is enforced by the unification algorithm. Furthermore, Extern hints can perform arbitrary computations and observe goals in many ways which break the equivalence used by the cache. So I think that rather than a cache, this should be dubbed a variant hint algorithm, which is never guaranteed to produce the same results as the original one.

It puts into the question the whole idea of typeclass caching. I am new to Coq codebase, so my responses to @ppedrot concerns are:

1. What are the bugs mentioned? Are these known bugs with tickets opened or yet undiscovered bugs? 
2. What is the case when the "result depends on fixed name"? I would love to see an example to anlyze how it interacts with my caching
3. The extern hints problem seems serious. Perhaps we can exlude from caching goals which were rejected using Extern hints?

What other people think? Is it the idea is dead in the water or we can still make it work?

**Comment by gares** [comment]
If we restrict caching to what I'd call "pure instances" (eg no hint extern, just terms) then I think the paper by Pientka I was mentioning should provide guidance about soundness and in which forms goals being proved have to be cached (what has to be fixed and what to be quantified). Here the paper: 
https://www.cs.mcgill.ca/~bpientka/papers/eftab-long.pdf for a reference.
The context is higher order logic programming, so the terminology is different. But TC resolution using pure instances is really logic programming, it is easy to make the link.

About performance, I agree with @ppedrot that doing things in a sound way is going to be expensive.
Still, blind search is (super) exponential, hence no matter how costly the cache is, it will always be a win in these cases. Of course one has to be able to tell the TC engine when to activate it (at least on a per-class and per-run basis). Or at least, this is what most logic programming languages do. If they provide tabling (not all of them do, that means it is not trivial to code) they let the user enable it on a specific predicate (ie class name).

My2C

**Comment by ejgallego** [comment]
Yeah, Enrico proposal to study the problem as tabling makes the most sense to me, and while tricky, it should be doable (at whatever cost) it does make sense to add tabling to the base LP search engine. Note that IMVHO this should be doable in a sound way, but it is going to be a very hard project (likeky on the level of producing a fully-fledged research paper)

**Comment by vzaliva** [comment]
Sorry for the delayed answer. I finally read the tabling paper. Few thoughts on it:

I do not see any immediate contradiction between their formalization and what we are doing.

However, the approach they describe is different as cache *successful* goals. Our approach of caching **failed** goals is significantly simpler. For example, it does not require producing proof witnesses, backtracking over multiple cached goals, keeping track of existential instantiations.

The particulars of their implementation (e.g. section 3.1) is optimized for their more genal use-case and an overkill for our approach.

The _linearization_ and tree-based indexing looks interesting and could be used as additional cache performance optimization. It is yet to be seen if it would give any tangible performance benefits as require complex goal manipulation and copying. Since current non-linearized lookup already shows performance improvements when the cache is used, these could be considered for later.

Getting back to proposed caching implementation:  

1. After reading the paper I still think our approach is sound, at least at the level used in the paper (high-order logic)
1. Some Coq/CIC specific things may pose additoinal challenges (e.g. hints, universes), but this tabling paper does not provide any additional insights into them.

I think our caching implementation is still useful, although a few additional problems need to be addressed:

1. Restrict caching to "pure instances" (eg no hint extern, just terms) 
2. Investigate further potential issue raised by @ppedrot regarding using global state during unification. There might or might not be a problem there.

I can work on these, but if there is a consensus that current approach is insufficient and only a more sophisticated approach (e.g. caching successful searches) will do I would rather abandon my attempt now if it does not have a chance making into production.

**Comment by Blaisorblade** [comment]
Just run into this issue. FWIW, a draft paper on TC caching in Lean 4 was released (as many already know).

**Comment by vzaliva** [comment]
Do you have a link to the paper? I would love to read it. I guess they have not tried to expose the buried Prolog interpreter in TC cache resolution. :)

**Comment by Blaisorblade** [comment]
https://arxiv.org/abs/2001.04301

**Comment by JasonGross** [comment]
It may also be worth having a within-single-execution cache that is sound and complete.  I think the following interface should be possible (@ppedrot, does this make sense to you?)
- The cache presents the interface of a single "lookup" function that takes an evar map, a context, and a goal, and returns a iterator for a lazily computed stream new-evar-map * typeclass solution.  (Perhaps even better would be to return a nested iterator, so that we can either say "give me the next solution which may differ only in the body of the proof term" or "give me the next solution which must differ in how evars or constraints are updated", so that we can cache this information too.)
- Two evar-map*goal*contexts are considered to be the same when they are syntactically equal differing only by a evar-universe-substitution (I expect this computation to be amortized constant overhead over just syntactic equality)

I think using this sort of cache should be enough for, e.g., https://github.com/coq/coq/issues/3921
